simplex/NonValidators.MD

```diff
@@ -0,0 +1,54 @@
+
+**Goal**: Allow nodes not participating in consensus (non-validators) to be able to replicate and verify accepted blocks.
+
+**Broadcasting Finalizations**
+
+In order to learn about blocks that were accepted, non-validators need to receive finalizations. To accommodate this, we should expand the `Communication` implementation to also broadcast to non-validator nodes whenever the message being broadcasted is a `FinalizationMessage`. This does not require any changes to the `Simplex` protocol itself, as the `Communication` component is already an isolated component. It would be up to the `Communication` struct and `Engine` that wraps `Simplex` to maintain a list of non-validator nodes and broadcast finalization messages to them.
+
+Once a non-validator receives a finalization message for a higher sequence, it can then request the full block via the standard replication path described below.
+
+**Standard Replication Path**: Requesting Missing Sequences
+When a validating node falls behind it will eventually receive a finalization for a `seq x` that is greater than its current known seq. A finalization for a higher sequence means the node has fallen behind, and blocks for sequences up to and including `seq x` are guaranteed to be accepted into the chain. In this case, the node will begin the replication process which involves requesting the sequences it is missing (all sequences from its last accepted seq to and including `x`). Eventually, the nodes will respond and the lagging node will have caught up.
+
+Simplex currently only accepts and processes messages from validators. These messages can be normal consensus messages or replication requests/responses.
+
+`ReplicationRequests` have the following structure:
+
+```go
+type ReplicationRequest struct {
+   Seqs        []uint64 // sequences we are requesting
+   LatestRound uint64   // latest round that we are aware of
+}
+
+```
+
+A `ReplicationResponse` has the following structure:
+
+```go
+type ReplicationResponse struct {
+   Data        []QuorumRound
+   LatestRound *QuorumRound
+}
+// QuorumRound represents a round that has achieved quorum on either
+// (empty notarization), (block & notarization), or (block, finalization)
+type QuorumRound struct {
+   Block             Block
+   Notarization      *Notarization
+   Finalization      *Finalization
+   EmptyNotarization *EmptyNotarization
+}
+```
+
+Non-validators should re-use this logic and can ignore any `QuorumRound` that do not have finalizations. The main benefit of reusing the same logic is that we can take advantage of the in-memory `rounds` map which will reduce the number of times we read from storage. Secondarily, this would be a minimalistic change that won't require maintaining a separate code path. In order to allow this, `Simplex` needs to be modified to accept `ReplicationRequests` from non-validator nodes.
```

Comment:
> In order to allow this, `Simplex` needs to be modified to accept `ReplicationRequests` from non-validator nodes.

I think the right approach would be to have a structure that will contain the epoch object and handle dispatching the requests instead of the epoch. That structure that contains the epoch inside of it, will reach into the rounds map of the epoch object if needed, to search for the missing block.

Reply:
I don't think that the Epoch object should dispatch requests from non-validators. As the code matures we would probably need to add some extra logic for nodes, such as throttling, or metrics, etc. and I think we should keep the responsibilities of the epoch object to the minimum to remove complexity.

Reply:
>  I can see it being useful to increase parallelization since we won't need to grab epoch lock, but were you thinking of something else?

If we would want to access the rounds map we would probably need to lock (maybe we'll want to make it an RWLock but would still need to lock)